### ISSUE TITLE
sieveFactor (factorSieve 2048) (29*73) == [(2117,1)]

### ISSUE THREAD

**Issue by CarlEdman** [cartazio/arithmoi#7]
This may just be a documentation bug or perhaps I am misreading this:

> sieveFactor fs n finds the prime factorisation of n using the FactorSieve fs. For negative n, a factor of -1 is included with multiplicity 1. After stripping any present factors 2, the remaining cofactor c (if larger than 1) is factorised with fs. This is most efficient of course if c does not exceed the bound with which fs was constructed. If it does, trial division is performed until either the cofactor falls below the bound or the sieve is exhausted. In the latter case, the elliptic curve method is used to finish the factorisation.

I understood this to mean that the FactorSieve is used to speed up the factorisation, but if an occasional argument to (sieveFactor fs) is higher than the argument to factorSieve, the results would still be correct (if less sped-up by the sieve).

However, this appears not to be the case.  For example with arithmoi 0.4.1.2 on Haskell Platform 2014.2 (i.e., ghc 7.8.3), I get, for example: '''sieveFactor (factorSieve 2048) (29*73)==[(2117,1)]'''

A some checks seem to find this discrepancy only in cases where the smallest prime factor is 29.

**Comment by CarlEdman** [comment]
Under 0.4.1.3 and ghc 7.10.2 still:
~~~
sieveFactor (factorSieve 2048) (29*73)==[(2117,1)]
~~~
Is this a real bug or am I misunderstanding something?

**Comment by cartazio** [comment]
I'll have to investigate myself , a lot of arithmoi is buggy / under tested
and has performance engineering that isn't validated with bench marks.

**Comment by cartazio** [comment]
If you can do any further sleuthing in the mean time that would be super
appreciated

**Comment by CarlEdman** [comment]
Actually, I discovered arithmoi over a year ago and when I benchmarked it against my own code, arithmoi just blew it away.  So since then, I've been using arithmoi for my own number theoretical investigations almost every day and that above bug is the only one I've found so far.

But by chance, I was just thinking about ways to improve arithmoi and, beside contributing a bunch of other standard number theoretic functions I've coded on top of arithmoi and which may be of interest to others, the biggest performance win I can see would be parallelisation of the code.  Many of the big number theory codes are embarrassingly parallelizable and I often see real-world 10x speed improvements on by 12-processor workstation just by adding a few judicious `par`s and `pseq`s.

**Comment by cartazio** [comment]
Cool! That's good to know, but we should still get benchmarks, especially
to make these performance ideas easier to compare and validate.  And that's
ignoring questions of various asymptotics etc :)

**Comment by Bodigrim** [comment]
@CarlEdman could you please test the following patch? It fixes issue with `2117`.

```
--- a/Math/NumberTheory/Primes/Sieve/Misc.hs
+++ b/Math/NumberTheory/Primes/Sieve/Misc.hs
@@ -149,7 +149,7 @@ sieveFactor (FS bnd sve) = check
           where
             p = toPrim ix
-            pix = unsafeAt sve ix
+            pix = unsafeAt sve $ fromIntegral p
     curve n = stdGenFactorisation (Just (bound*(bound+2))) (mkStdGen $ fromIntegral n `xor` 0xdecaf00d) Nothing n
```

**Comment by CarlEdman** [comment]
> @CarlEdman could you please test the following patch? It fixes issue with 2117.

Thanks!  That fixes the issue with my test case.

**Comment by Bodigrim** [comment]
Alas, we still have no tests of `Math.NumberTheory.Primes`. It would be nice to have any, even not exhaustive ones. 

"contributing a bunch of other standard number theoretic functions" (accompanied with tests) would be great.

I am not sure that low-level libraries (and `arithmoi` definitely is low level) should use play with parallelism. Maybe we can put such high-level stuff in a separate package.

**Comment by CarlEdman** [comment]
Exhaustive tests of the prime functions might not be feasible with finite time and space. :)  That said, throwing together a quickCheck which takes some random numbers, factorises, multiplies, factorises the product, and then tests each of the factors for primality shouldn't be too hard.  Of course, I have not written such a thing.

Ok, going through my personal library and checking for things which might be of general interest, I have pisanoPeriod, pythTrips, pythQuads, inverseTotient, integerPartitions, periods, repsTwoSquaresNo, repsTwoSquares, moebius, primitiveRoots, groupOrder, discreteLog, and moebiusDivisors.  Most of them are based on the fast arithmoi routines.

And I agree that arithmoi is low-level, but is that a compelling argument not to do parallelism?  As far as I understand the algorithms, they ought to be highly parallelizable without changing the API or non-trivially affecting single-threaded performance.  Give how CPU intensive they are and how difficult effective parallelism can be, I think this could be a huge win (as I mentioned, real-world 1,000% speedups on my 12 HEC machine).

**Comment by Bodigrim** [comment]
By "exhaustive" I mean tests which trigger each expression of the source code at least once. This is quite feasible, however difficult.

The performance of parallelism heavily depends on the amount of work for each thread. If tasks are too small, then the paralleled computation could easily be slower than a single-thread one due to overhead on GC and sync.

If we want to factorise a really large number, it may be great to have a parallel version of `factorise`. But if we need to factorise a list of middle-sized integers, we'd better run each factorisation in a separate single thread. And if the list is really long and numbers are small, we should split it into chunks (e.g., 1000 elements per chunk) and divide them among threads. IMHO there is no way to determine the desirable level of parallelism at the level of library routines.

Regarding your personal library - the list of the functions sounds exciting. Could you please make a pull request or (if it is easier) publish them as a gist?

**Comment by Bodigrim** [comment]
The issue was fixed in #17.
